docs/src/posts/configuration.md
---
title: Configuration
---

## Defaults

These are the default options from `nvim-dap-view`. You can use them as reference. You don't have to copy-paste them.

```lua
return {
    winbar = {
        show = true,
        -- You can add a "console" section to merge the terminal with the other views
        sections = { "watches", "scopes", "exceptions", "breakpoints", "threads", "repl" },
        -- Must be one of the sections declared above (except for "console")
        default_section = "watches",
        -- Configure each section individually
        base_sections = {
            breakpoints = {
                keymap = "B",
                label = "Breakpoints [B]",
                short_label = " [B]",
                action = function()
                    require("dap-view.views").switch_to_view("breakpoints")
                end,
            },
            scopes = {
                keymap = "S",
                label = "Scopes [S]",
                short_label = "󰂥 [S]",
                action = function()
                    require("dap-view.views").switch_to_view("scopes")
                end,
            },
            exceptions = {
                keymap = "E",
                label = "Exceptions [E]",
                short_label = "󰢃 [E]",
                action = function()
                    require("dap-view.views").switch_to_view("exceptions")
                end,
            },
            watches = {
                keymap = "W",
                label = "Watches [W]",
                short_label = "󰛐 [W]",
                action = function()
                    require("dap-view.views").switch_to_view("watches")
                end,
            },
            threads = {
                keymap = "T",
                label = "Threads [T]",
                short_label = "󱉯 [T]",
                action = function()
                    require("dap-view.views").switch_to_view("threads")
                end,
            },
            repl = {
                keymap = "R",
                label = "REPL [R]",
                short_label = "󰯃 [R]",
                action = function()
                    require("dap-view.repl").show()
                end,
            },
            console = {
                keymap = "C",
                label = "Console [C]",
                short_label = "󰆍 [C]",
                action = function()
                    require("dap-view.term").show()
                end,
            },
        },
        -- Add your own sections
        custom_sections = {},
        controls = {
            enabled = false,
            position = "right",
            buttons = {
                "play",
                "step_into",
                "step_over",
                "step_out",
                "step_back",
                "run_last",
                "terminate",
                "disconnect",
            },
            icons = {
                play = "",
                pause = "",
                step_into = "",
                step_over = "",
                step_out = "",
                step_back = "",
                run_last = "",
                terminate = "",
                disconnect = "",
            },
            custom_buttons = {},
        },
    },
    windows = {
        height = 0.25,
        position = "below",
        terminal = {
            width = 0.5,
            position = "left",
            -- List of debug adapters for which the terminal should be ALWAYS hidden
            hide = {},
            -- Hide the terminal when starting a new session
            start_hidden = false,
        },
    },
    help = {
        border = nil,
    },
    -- Controls how to jump when selecting a breakpoint or navigating the stack
    switchbuf = "usetab",
    auto_toggle = false,
}
```


CHANGELOG.md
# Changelog

## 1.0.0 (2025-07-30)


### ⚠ BREAKING CHANGES

* **buttons:** restore previous config for icons
* **controls:** full customization of default buttons
* make default height proportional to screen size
* full customization of sections
* do not use `newtab` by default for switchbuf
* **term:** delete buffers when finishing sessions
* cleanup README
* remove help file
* do not autogenerate help file
* **threads:** only show valid frames
* specify adapters for which the terminal should be hidden ([#10](https://github.com/igorlfs/nvim-dap-view/issues/10))
* avoid boolean concat
* remove hooks that close the plugin on session end
* swapped `show`

### fixup

* avoid boolean concat ([3b79d9d](https://github.com/igorlfs/nvim-dap-view/commit/3b79d9de31c45f461e0e39570b176113aeb93e2c))
* swapped `show` ([08ad351](https://github.com/igorlfs/nvim-dap-view/commit/08ad351aecb76e3f6afb4db3e04d2fc835397ced))


### Features

* `DapViewShow` command ([c31ef8b](https://github.com/igorlfs/nvim-dap-view/commit/c31ef8b67e3dfd1751957f8cb832cbc37f6e478e))
* `DapViewToggle!`, `DapViewClose!`, start_hidden option ([#17](https://github.com/igorlfs/nvim-dap-view/issues/17)) ([c86697c](https://github.com/igorlfs/nvim-dap-view/commit/c86697ce5ee5bd7930f3ec3895b8a7b01a98bc52))
* adds filetypes to dap-view buffers ([5d8daeb](https://github.com/igorlfs/nvim-dap-view/commit/5d8daeb5c9d4922914c2fde3a386ebe6371be701))
* allow collapsing via foldmethod = indent ([ca7906a](https://github.com/igorlfs/nvim-dap-view/commit/ca7906af3c787c8952f87c1ed58c405341fa3d5d))
* allow custom sections to set buffer function and ft ([902e8ed](https://github.com/igorlfs/nvim-dap-view/commit/902e8ed4fe632fdc46503484ec5a6b4a4419c3a4))
* allow evaluating more complex expressions ([ba72154](https://github.com/igorlfs/nvim-dap-view/commit/ba721546577e1c16bbb24e252ee62693e57790d4))
* allow extensions to register views ([ad85982](https://github.com/igorlfs/nvim-dap-view/commit/ad859822e1bd144738cd57d0ca168d40197c986f))
* allow height to be float ([0d537ca](https://github.com/igorlfs/nvim-dap-view/commit/0d537ca353a225496b9e8ab0063d9f44275a8ff0))
* allow overriding help border ([b0cd4e1](https://github.com/igorlfs/nvim-dap-view/commit/b0cd4e1eb7b856e4f7359923fdf4397324cd92ac))
* allow spawning windows in a different direction ([5e69020](https://github.com/igorlfs/nvim-dap-view/commit/5e69020bba51c790d3daa77921f2382850b46cad))
* basic config validation ([28d0d54](https://github.com/igorlfs/nvim-dap-view/commit/28d0d542caaddf38e3cfbd9c295db35ad9ff37fe))
* **breakpoints:** delete breakpoint ([93d44eb](https://github.com/igorlfs/nvim-dap-view/commit/93d44eb7624187f49ed7d15a0948bfc38591116f))
* **breakpoints:** jump to a breakpoint ([90da63c](https://github.com/igorlfs/nvim-dap-view/commit/90da63caa5fedc9e5571bebc7934980eeed16ab1))
* **buttons:** restore previous config for icons ([390dae6](https://github.com/igorlfs/nvim-dap-view/commit/390dae6bf67f3342ebb481159932ef0fe54822ba))
* close term window if outside of session ([17ec283](https://github.com/igorlfs/nvim-dap-view/commit/17ec2831dd94b06188cd442c31865ce35123364b))
* colors, handle deleting expressions ([ca7906a](https://github.com/igorlfs/nvim-dap-view/commit/ca7906af3c787c8952f87c1ed58c405341fa3d5d))
* **config:** allow hiding winbar ([3eddba5](https://github.com/igorlfs/nvim-dap-view/commit/3eddba5b5b0879e6e1e5720c1b2e57567db3e015))
* configurable height ([84dbbcf](https://github.com/igorlfs/nvim-dap-view/commit/84dbbcf79dd5576da2eef14b17d76e7fd86e74b0))
* console view ([589dd18](https://github.com/igorlfs/nvim-dap-view/commit/589dd1802137289b3ffbfd8448b79d685dcd7cdf))
* console view ([7d6335e](https://github.com/igorlfs/nvim-dap-view/commit/7d6335ea5f5f4699f9d367b4a95131e6ad6bcd07))
* **controls:** full customization of default buttons ([25dead9](https://github.com/igorlfs/nvim-dap-view/commit/25dead9fc459ca8a1e555dda74fcfea711917290))
* copy expression to clipboard ([ca7906a](https://github.com/igorlfs/nvim-dap-view/commit/ca7906af3c787c8952f87c1ed58c405341fa3d5d))
* DapViewJump command ([cd8781a](https://github.com/igorlfs/nvim-dap-view/commit/cd8781a0a63c1384c12839a22c2d6a8571cc241d))
* define custom hl groups ([29f6e7a](https://github.com/igorlfs/nvim-dap-view/commit/29f6e7a16c014423d55951017efe8ff9957aa202))
* delete watch expr ([1fbebae](https://github.com/igorlfs/nvim-dap-view/commit/1fbebaeb3badc760928cf5cf79a2949bb64f3160))
* edit expression ([fba18ea](https://github.com/igorlfs/nvim-dap-view/commit/fba18eae850740e9f21b65465855fc96657e6c58))
* edit expressions ([ca7906a](https://github.com/igorlfs/nvim-dap-view/commit/ca7906af3c787c8952f87c1ed58c405341fa3d5d))
* **exceptions:** store exceptions_options per adapter ([2c733c0](https://github.com/igorlfs/nvim-dap-view/commit/2c733c0570952f9edfd7f74ceba18538d7d06f80))
* experimental REPL support ([49ce20e](https://github.com/igorlfs/nvim-dap-view/commit/49ce20eef5226cc40f769076ed7cba96a9e9bf2e))
* expose terminal bufnr/winnr in state module ([#19](https://github.com/igorlfs/nvim-dap-view/issues/19)) ([a0c5f9f](https://github.com/igorlfs/nvim-dap-view/commit/a0c5f9f59a3b65397c01de12722ecf1eae00a52e))
* expose user commands ([6695fd1](https://github.com/igorlfs/nvim-dap-view/commit/6695fd1ad42fbbd2d7ffdb90042ccc5ad2e3a2f3))
* full customization of sections ([108c05a](https://github.com/igorlfs/nvim-dap-view/commit/108c05a85bd49257436dda728041b8099caf68f7))
* handle emptiness in views ([a59639a](https://github.com/igorlfs/nvim-dap-view/commit/a59639ae46ef4d9feed721d675bfc0d6982eaa1e))
* highlight variables that were updated with a different color ([4fb6235](https://github.com/igorlfs/nvim-dap-view/commit/4fb62357ce36554526060d470cf27135eb297bf6))
* hijack terminal ([ab13019](https://github.com/igorlfs/nvim-dap-view/commit/ab13019b9d3705d7e97c7fcf02dbbc528ec5b297))
* improve multitab support ([960e219](https://github.com/igorlfs/nvim-dap-view/commit/960e219c6a12b4cd60413d7918c877a1454cee31))

lua/dap-view/config.lua
local views = require("dap-view.views")

local M = {}

---@alias dapview.CustomSection string
---@alias dapview.DefaultSection "breakpoints" | "exceptions" | "watches" | "repl" | "threads" | "console" | "scopes"
---@alias dapview.Section dapview.DefaultSection | dapview.CustomSection

---@alias dapview.CustomButton string
---@alias dapview.DefaultButton "play" | "step_into" | "step_over" | "step_out" | "step_back" | "run_last" | "terminate" | "disconnect"
---@alias dapview.Button dapview.CustomButton | dapview.DefaultButton

---@class dapview.TerminalConfig
---@field hide string[] List of adapters for which the terminal should be hidden
---@field position 'right' | 'left' | 'above' | 'below'
---@field width number If > 1 number of columns, else percentage the terminal window should use
---@field start_hidden boolean Don't show the terminal window when starting a new session

---@class dapview.WindowsConfig
---@field height number If > 1 number of lines, else percentage the windows should use
---@field position 'right' | 'left' | 'above' | 'below'
---@field anchor? fun(): integer? Function that returns a window number for the main nvim-dap-view window to follow
---@field terminal dapview.TerminalConfig

---@alias dapview.DefaultIcons dapview.DefaultButton | "pause"

---@class dapview.ControlsIconsConfig
---@field pause string
---@field play string
---@field step_into string
---@field step_over string
---@field step_out string
---@field step_back string
---@field run_last string
---@field terminate string
---@field disconnect string

---@class dapview.ButtonConfig
---@field render fun(session?: dap.Session): string Render the button (highlight and icon). Receives the current session as a param.
---@field action fun(clicks: integer, button: string, modifiers: string): nil Click handler. See `:help statusline`

---@class dapview.ControlsConfig
---@field enabled boolean
---@field position 'left' | 'right'
---@field buttons dapview.Button[] Buttons to show in the controls section
---@field custom_buttons table<dapview.CustomButton, dapview.ButtonConfig> Custom buttons to show in the controls section
---@field icons dapview.ControlsIconsConfig Icons for each button

---@class dapview.SectionConfig
---@field label string
---@field short_label string Label to be shown if there's not enough space to display the entire winbar
---@field keymap string
---@field action fun(): nil

---@class dapview.CustomSectionConfig : dapview.SectionConfig
---@field buffer fun(): integer Creates a new buffer for the section
---@field filetype string Filetype used by the section

---@class dapview.WinbarConfig
---@field sections dapview.Section[]
---@field default_section dapview.Section
---@field show boolean
---@field base_sections table<dapview.Section,dapview.SectionConfig>
---@field custom_sections table<dapview.CustomSection, dapview.CustomSectionConfig>
---@field controls dapview.ControlsConfig

---@class dapview.HelpConfig
---@field border? string|string[] Override `winborder`

---@class (exact) dapview.ConfigStrict
---@field winbar dapview.WinbarConfig
---@field windows dapview.WindowsConfig
---@field help dapview.HelpConfig
---@field switchbuf string Control how to jump when selecting a breakpoint or a call in the stack
---@field auto_toggle boolean

---@type dapview.ConfigStrict
M.config = {
    winbar = {
        show = true,
        sections = { "watches", "scopes", "exceptions", "breakpoints", "threads", "repl" },
        default_section = "watches",
        base_sections = {
            breakpoints = {
                keymap = "B",
                label = "Breakpoints [B]",
                short_label = " [B]",
                action = function()
                    views.switch_to_view("breakpoints")
                end,
            },
            scopes = {
                keymap = "S",
                label = "Scopes [S]",
                short_label = "󰂥 [S]",
                action = function()
                    views.switch_to_view("scopes")
                end,
            },
            exceptions = {
                keymap = "E",
                label = "Exceptions [E]",
                short_label = "󰢃 [E]",
                action = function()
                    views.switch_to_view("exceptions")
                end,
            },
            watches = {
                keymap = "W",
                label = "Watches [W]",
                short_label = "󰛐 [W]",
                action = function()
                    views.switch_to_view("watches")
                end,
            },
            threads = {
                keymap = "T",
                label = "Threads [T]",
                short_label = "󱉯 [T]",
                action = function()
                    views.switch_to_view("threads")
                end,
            },
            repl = {
                keymap = "R",
                label = "REPL [R]",
                short_label = "󰯃 [R]",
                action = function()
                    require("dap-view.repl").show()
                end,
            },
            console = {
                keymap = "C",
                label = "Console [C]",
                short_label = "󰆍 [C]",
                action = function()
                    require("dap-view.term").show()
                end,
            },
        },
        custom_sections = {},
        controls = {
            enabled = false,
            position = "right",
            buttons = {
                "play",
                "step_into",
                "step_over",
                "step_out",
                "step_back",
                "run_last",
                "terminate",
                "disconnect",
            },
            icons = {
                play = "",
                pause = "",
                step_into = "",
                step_over = "",
                step_out = "",
                step_back = "",
                run_last = "",
                terminate = "",
                disconnect = "",
            },
            custom_buttons = {},
        },
    },
    windows = {
        height = 0.25,
        position = "below",
        terminal = {
            width = 0.5,
            position = "left",
            hide = {},
            start_hidden = false,
        },
    },
    help = {
        border = nil,
    },
    switchbuf = "usetab",
    auto_toggle = false,
}

return M
